File: android/app/src/main/kotlin/com/example/kimoi/extractors/OkruExtractor.kt
package com.example.kimoi.extractors

import org.jsoup.nodes.Document
import com.example.kimoi.model.Video
import java.net.HttpURLConnection
import java.net.URL

class OkruExtractor() {

    private fun fixQuality(quality: String): String {
        val qualities = listOf(
            Pair("ultra", "2160p"),
            Pair("quad", "1440p"),
            Pair("full", "1080p"),
            Pair("hd", "720p"),
            Pair("sd", "480p"),
            Pair("low", "360p"),
            Pair("lowest", "240p"),
            Pair("mobile", "144p"),
        )
        return qualities.find { it.first == quality }?.second ?: quality
    }

    fun videosFromUrl(url: String, prefix: String = "", fixQualities: Boolean = true): List<Video> {

        val myURL = URL(url)
        val connection = myURL.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.setRequestProperty("Accept", "application/json")
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36")
        connection.connect()

        val inputStream = connection.inputStream
        val response = inputStream.bufferedReader().readText()

        val document = Document(url).parser().parseInput(response, url)


        val videosString = document.selectFirst("div[data-options]")
            ?.attr("data-options")
            ?.substringAfter("\\\"videos\\\":[{\\\"name\\\":\\\"")
            ?.substringBefore("]")
            ?: return emptyList<Video>()
        return videosString.split("{\\\"name\\\":\\\"").reversed().mapNotNull {
            val videoUrl = it.substringAfter("url\\\":\\\"")
                .substringBefore("\\\"")
                .replace("\\\\u0026", "&")
            val quality = it.substringBefore("\\\"").let {
                if (fixQualities) {
                    fixQuality(it)
                } else {
                    it
                }
            }
            val videoQuality = ("Okru:" + quality).let {
                if (prefix.isNotBlank()) {
                    "$prefix $it"
                } else {
                    it
                }
            }
            if (videoUrl.startsWith("https://")) {
                Video(videoUrl, videoQuality, videoUrl, server = "Okru")
            } else {
                null
            }
        }
    }
}

File: lib/src/utils/extractors/okru_extractor.dart
import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';
import 'package:html/parser.dart';
import 'package:http/http.dart';
import 'package:kimoi/src/infrastructure/infrastructure.dart';
import 'package:kimoi/src/utils/extensions/extension.dart';

class OkruExtractor {
  final client = Client();

  final headers = {
    "User-Agent":
        "Mozilla/5.0 (Linux; Android 5.1.1; Android SDK built for x86 Build/LMY48X) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/39.0.0.0 Mobile Safari/537.36"
  };

  String fixQuality(String quality) {
    final qualities = {
      "ultra": "2160p",
      "quad": "1440p",
      "full": "1080p",
      "hd": "720p",
      "sd": "480p",
      "low": "360p",
      "lowest": "240p",
      "mobile": "144p"
    };

    return qualities[quality] ?? 'Okru';
  }

  Future<List<Video>> videosFromUrl(String url) async {
    String? html;

    bool isLoading = true;

    try {
      HeadlessInAppWebView(
        initialUrlRequest: URLRequest(url: Uri.parse(url), headers: headers),
        
        onLoadStop: (controller, url) async {
          html = await controller.getHtml();
          isLoading = false;
        },
      )
        ..run()
        ..dispose();

      while (isLoading) {
        await Future.delayed(const Duration(milliseconds: 500));
      }

      var doc = parse(html);

      final data =
          doc.querySelector('div[data-options]')?.attributes['data-options'];

      if (data == null) return [];

      // if (data.contains("ondemandHls")) {
      //   print("ondemandHls");
      // } else if (data.contains("ondemandHls")) {
      //   print("ondemandHls");
      // }

      final videos = videosFromJson(data);
      return videos;
    } on Exception catch (e) {
      debugPrint('$e');
      return [];
    }
  }

  List<Video> videosFromJson(String videoString,
      {String prefix = "", bool fixQualities = true}) {
    final arrayData = videoString
        .substringAfter("\\\"videos\\\":[{\\\"name\\\":\\\"")
        .substringBefore("]");

    final data = arrayData.split("{\\\"name\\\":\\\"").reversed.map((it) {
      final videoUrl = it.extractLink("url");
      final resolution = it.substringBefore("\\\"");
      final quality = fixQuality(resolution);

      final videoQuality = "Okru:$quality";

      // if (videoUrl.startsWith("https://")) {
      return Video(
          url: videoUrl,
          quality: videoQuality,
          videoUrl: videoUrl,
          headers: headers);
      // } else {
      //   null;
      // }
    }).toList();

    return data;
  }
}
